<a href="https://colab.research.google.com/github/sashapisterman/DSO510-Business-Analytics/blob/main/Assignment_1/DSO_510_Assignment_1_Pisterman_Sasha.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Week 1 Assignment
## Python Foundations and Basic Data Analysis: FoodHub Data Analysis

**Total: 30 points**

### Context

The number of restaurants in New York is increasing day by day. Lots of students and busy professionals rely on those restaurants due to their hectic lifestyles. Online food delivery service is a great option for them. It provides them with good food from their favorite restaurants. A food aggregator company FoodHub offers access to multiple restaurants through a single smartphone app.

The app allows the restaurants to receive a direct online order from a customer. The app assigns a delivery person from the company to pick up the order after it is confirmed by the restaurant. The delivery person then uses the map to reach the restaurant and waits for the food package. Once the food package is handed over to the delivery person, he/she confirms the pick-up in the app and travels to the customer's location to deliver the food. The delivery person confirms the drop-off in the app after delivering the food package to the customer. The customer can rate the order in the app. The food aggregator earns money by collecting a fixed margin of the delivery order from the restaurants.

### Objective

The food aggregator company has stored the data of the different orders made by the registered customers in their online portal. They want to analyze the data to get a fair idea about the demand of different restaurants which will help them in enhancing their customer experience. Suppose you are a Data Scientist at Foodhub and the Data Science team has shared some of the key questions that need to be answered. Perform the data analysis to find answers to these questions that will help the company to improve the business.

### Data Description

The data contains the different data related to a food order. The detailed data dictionary is given below.

### Data Dictionary

* order_id: Unique ID of the order
* customer_id: ID of the customer who ordered the food
* restaurant_name: Name of the restaurant
* cuisine_type: Cuisine ordered by the customer
* cost_of_the_order: Cost of the order
* day_of_the_week: Indicates whether the order is placed on a weekday or weekend (The weekday is from Monday to Friday and the weekend is Saturday and Sunday)
* rating: Rating given by the customer out of 5
* food_preparation_time: Time (in minutes) taken by the restaurant to prepare the food. This is calculated by taking the difference between the timestamps of the restaurant's order confirmation and the delivery person's pick-up confirmation.
* delivery_time: Time (in minutes) taken by the delivery person to deliver the food package. This is calculated by taking the difference between the timestamps of the delivery person's pick-up confirmation and drop-off information

### **Please read the instructions carefully before starting the project.**
This is a commented Jupyter IPython Notebook file in which all the instructions and tasks to be performed are mentioned. Read along carefully to complete the project.
* Blanks '_______' are provided in the notebook that
needs to be filled with an appropriate code to get the correct result. Please replace the blank with the right code snippet. With every '_______' blank, there is a comment that briefly describes what needs to be filled in the blank space.
* Identify the task to be performed correctly, and only then proceed to write the required code.
* Fill the code wherever asked by the commented lines like "# write your code here" or "# complete the code". Running incomplete code may throw an error.
* Please run the codes in a sequential manner from the beginning to avoid any unnecessary errors.

### Let us start by importing the required libraries

In [1]:
# Import libraries for data manipulation
import numpy as np
import pandas as pd

### Understanding the structure of the data

In [2]:
# uncomment and run the following lines for Google Colab
# from google.colab import drive
# drive.mount('/content/drive')

In [3]:
# uncomment and run the following lines for importing file from computer

#from google.colab import files
#uploaded = files.upload()

### **Question 0:** Import the data into Colab environment? [1 point]

In [4]:
# Loading the FoodHub order dataset into a pandas DataFrame
df = pd.read_csv('foodhub_order.csv')
# Displaying the first 5 rows to verify the data loaded correctly
df.head()

,order_id,customer_id,restaurant_name,cuisine_type,cost_of_the_order,day_of_the_week,rating,food_preparation_time,delivery_time
0,1477147,337525,Hangawi,Korean,30.75,Weekend,Not given,25,20
1,1477685,358141,Blue Ribbon Sushi Izakaya,Japanese,12.08,Weekend,Not given,25,23
2,1477070,66393,Cafe Habana,Mexican,12.23,Weekday,5,23,28
3,1477334,106968,Blue Ribbon Fried Chicken,American,29.20,Weekend,3,25,15
4,1478249,76942,Dirty Bird to Go,American,11.59,Weekday,4,25,24


### **Question 1:** How many rows and columns are present in the data? [1 point]

In [5]:
# Checking the number of rows and columns in the dataset
df.shape

(1898, 9)

### **Question 2:** What are the datatypes of the different columns in the dataset? [1 point]

In [6]:
# Checking the datatype of each column in the dataset
df.dtypes

,0
order_id,int64
customer_id,int64
restaurant_name,object
cuisine_type,object
cost_of_the_order,float64
day_of_the_week,object
rating,object
food_preparation_time,int64
delivery_time,int64


### **Question 3:** Are there any missing values in the data? If yes, treat them using an appropriate method.  [1 point]

In [7]:
# Checking for missing values in each column
df.isnull().sum()

,0
order_id,0
customer_id,0
restaurant_name,0
cuisine_type,0
cost_of_the_order,0
day_of_the_week,0
rating,0
food_preparation_time,0
delivery_time,0


The dataset has no missing values, so no treatment is required.

### **Question 4:** Check the statistical summary of the data. What is the minimum, average, and maximum time it takes for food to be prepared once an order is placed? [2 points]

In [8]:
# Getting the statistical summary of the numerical columns
df.describe()

,order_id,customer_id,cost_of_the_order,food_preparation_time,delivery_time
count,1.898000e+03,1898.000000,1898.000000,1898.000000,1898.000000
mean,1.477496e+06,171168.478398,16.498851,27.371970,24.161749
std,5.480497e+02,113698.139743,7.483812,4.632481,4.972637
min,1.476547e+06,1311.000000,4.470000,20.000000,15.000000
25%,1.477021e+06,77787.750000,12.080000,23.000000,20.000000
50%,1.477496e+06,128600.000000,14.140000,27.000000,25.000000
75%,1.477970e+06,270525.000000,22.297500,31.000000,28.000000
max,1.478444e+06,405334.000000,35.410000,35.000000,33.000000


The food preparation time ranges from a minimum of 20 minutes to a
maximum of 35 minutes, with an average of approximately 27.37 minutes.
This suggests that most orders take around 27 minutes to prepare, and
preparation time does not vary drastically across orders.

### **Question 5:** How many orders are not rated? [1 point]

In [9]:
# Checking the distribution of ratings to find unrated orders
df['rating'].value_counts()

,count
rating,
Not given,736
5,588
4,386
3,188


736 out of 1,898 orders were not rated, which represents approximately
38.8% of all orders. This is a significant portion of unrated orders
and may suggest that the app should do more to encourage customers to
leave ratings after their delivery.

### **Question 6:** Which are the top 5 restaurants in terms of the number of orders received? [1 point]

In [10]:
# Get top 5 restaurants with highest number of orders
df['restaurant_name'].value_counts().head()

,count
restaurant_name,
Shake Shack,219
The Meatball Shop,132
Blue Ribbon Sushi,119
Blue Ribbon Fried Chicken,96
Parm,68


The top 5 restaurants by number of orders are Shake Shack (219),
The Meatball Shop (132), Blue Ribbon Sushi (119), Blue Ribbon Fried
Chicken (96), and Parm (68). Shake Shack leads by a significant margin
with nearly double the orders of the second place restaurant, suggesting
it is by far the most popular restaurant on the platform. The Blue Ribbon
brand appears twice in the top 5, indicating strong overall brand
performance on FoodHub.

### **Question 7:** Which is the most popular cuisine on weekends? [2 points]

In [11]:
# Filtering the dataset to only include weekend orders
df_weekend = df[df['day_of_the_week'] == 'Weekend']
# Checking the most popular cuisine type on weekends
df_weekend['cuisine_type'].value_counts()

,count
cuisine_type,
American,415
Japanese,335
Italian,207
Chinese,163
Mexican,53
Indian,49
Middle Eastern,32
Mediterranean,32
Thai,15


On weekends, American cuisine is the most popular with 415 orders,
followed by Japanese (335), Italian (207), Chinese (163), and Mexican
(53). American cuisine accounts for a significant portion of weekend
orders, suggesting that FoodHub should ensure strong representation of
American restaurants on the platform, particularly on weekends.

### **Question 8:** What percentage of the orders cost more than 20 dollars? [2 points]


In [12]:
# Filtering orders that cost more than $20
df_greater_than_20 = df[df['cost_of_the_order']>20]

# Calculate the number of total orders where the cost is above 20 dollars
print('The number of total orders that cost above 20 dollars is:', df_greater_than_20.shape[0])

# Calculate percentage of such orders in the dataset
percentage = (df_greater_than_20.shape[0] / df.shape[0]) * 100

print("Percentage of orders above 20 dollars:", round(percentage, 2), '%')

The number of total orders that cost above 20 dollars is: 555
Percentage of orders above 20 dollars: 29.24 %


555 out of 1,898 orders cost more than 20 dollars, representing approximately
29.24% of all orders. This means that nearly 1 in 3 orders exceeds 20 dollars,
suggesting that FoodHub customers are willing to spend on higher priced meals,
which is a positive indicator for revenue generation.

### **Question 9:** What is the mean order delivery time? [1 point]

In [13]:
# Calculating the mean delivery time
mean_del_time = df['delivery_time'].mean()
print('The mean delivery time for this dataset is', round(mean_del_time, 2), 'minutes')

The mean delivery time for this dataset is 24.16 minutes


The mean delivery time across all orders is 24.16 minutes. Combined
with the average food preparation time of 27.37 minutes, the average
total time a customer waits from placing an order to receiving it is
approximately 51.53 minutes, which is just under an hour.

### **Question 10:** The company has decided to give 20% discount vouchers to the top 3 most frequent customers. Find the IDs of these customers and the number of orders they placed. [1 point]

In [14]:
# Getting the top 3 most frequent customers for discount vouchers
df['customer_id'].value_counts().head(3)

,count
customer_id,
52832,13
47440,10
83287,9


The top 3 most frequent customers are customer 52832 with 13 orders,
customer 47440 with 10 orders, and customer 83287 with 9 orders.
These customers should receive the 20% discount vouchers. Their high
order frequency suggests strong loyalty to the platform, and the
discount voucher is a good strategy to retain them.

### **Question 11:** The company wants to provide a promotional offer in the advertisement of the restaurants. The condition to get the offer is that the restaurants must have a rating count of more than 50 and the average rating should be greater than 4. Find the restaurants fulfilling the criteria to get the promotional offer. [3 points]

In [15]:
# Filter the rated restaurants
df_rated = df[df['rating'] != 'Not given'].copy()

# Convert rating column from object to integer
df_rated['rating'] = df_rated['rating'].astype('int')

# Create a dataframe that contains the restaurant names with their rating counts
df_rating_count = df_rated.groupby(['restaurant_name'])['rating'].count().sort_values(ascending = False).reset_index()
df_rating_count.head()

,restaurant_name,rating
0,Shake Shack,133
1,The Meatball Shop,84
2,Blue Ribbon Sushi,73
3,Blue Ribbon Fried Chicken,64
4,RedFarm Broadway,41


In [16]:
# Getting restaurant names with more than 50 ratings
rest_names = df_rating_count[df_rating_count['rating'] > 50]['restaurant_name']

# Filtering to get only those restaurants
df_mean_4 = df_rated[df_rated['restaurant_name'].isin(rest_names)].copy()

# Finding restaurants with mean rating greater than 4
df_mean_4.groupby(['restaurant_name'])['rating'].mean().sort_values(ascending = False).reset_index().dropna() ## Complete the code to find the mean rating


,restaurant_name,rating
0,The Meatball Shop,4.511905
1,Blue Ribbon Fried Chicken,4.328125
2,Shake Shack,4.278195
3,Blue Ribbon Sushi,4.219178


Four restaurants meet the criteria for the promotional offer, having
both a rating count of more than 50 and an average rating greater than 4:
The Meatball Shop, Blue Ribbon Fried Chicken, Shake Shack, and Blue Ribbon
Sushi. Notably, three of these four restaurants also appeared in the top 5
most ordered restaurants from Question 6, suggesting that higher order
volume and higher customer satisfaction tend to go hand in hand on the
FoodHub platform.

### **Question 12:** The company charges the restaurant 25% on the orders having cost greater than 20 dollars and 15% on the orders having cost greater than 5 dollars. Find the net revenue generated by the company across all orders. [3 marks]

In [17]:
#Function to determine the revenue based on order cost
def compute_rev(x):
    if x > 20:
        return x*0.25
    elif x > 5:
        return x*0.15
    else:
        return x*0

# Applying the revenue function to the cost of the order column
df['Revenue'] = df['cost_of_the_order'].apply(compute_rev)
df.head()

,order_id,customer_id,restaurant_name,cuisine_type,cost_of_the_order,day_of_the_week,rating,food_preparation_time,delivery_time,Revenue
0,1477147,337525,Hangawi,Korean,30.75,Weekend,Not given,25,20,7.6875
1,1477685,358141,Blue Ribbon Sushi Izakaya,Japanese,12.08,Weekend,Not given,25,23,1.8120
2,1477070,66393,Cafe Habana,Mexican,12.23,Weekday,5,23,28,1.8345
3,1477334,106968,Blue Ribbon Fried Chicken,American,29.20,Weekend,3,25,15,7.3000
4,1478249,76942,Dirty Bird to Go,American,11.59,Weekday,4,25,24,1.7385


In [18]:
# Calculating the total net revenue across all orders
total_rev = df['Revenue'].sum()
print('The net revenue is around', round(total_rev, 2), 'dollars')

The net revenue is around 6166.3 dollars


The net revenue generated by FoodHub across all orders is 6,166.30 dollars.
This revenue is collected by charging 25% on orders costing more than 20 dollars
and 15% on orders costing more than 5 dollars. Given that there are 1,898 total
orders, this works out to an average revenue of approximately 3.25 dollars per
order, suggesting there may be an opportunity to review the commission structure
to increase revenue per order.

### **Question 13:** The company wants to analyze the total time required to deliver the food. What percentage of orders take more than 60 minutes to get delivered from the time the order is placed? (The food has to be prepared and then delivered.)[3 points]


In [19]:
# Calculating total delivery time by adding food preparation and delivery time
df['total_time'] = df['food_preparation_time'] + df['delivery_time']

# Filtering orders that take more than 60 minutes
df_greater_than_60 = df[df['total_time'] > 60]

# Calculating the percentage of orders that take more than 60 minutes
percentage_60 = (df_greater_than_60.shape[0] / df.shape[0]) * 100

print('Percentage of orders that take more than 60 minutes:', round(percentage_60, 2), '%')

Percentage of orders that take more than 60 minutes: 10.54 %


Approximately 10.54% of all orders take more than 60 minutes to be
delivered from the time the order is placed. While this represents
a relatively small portion of orders, it still affects roughly 1 in
10 customers, which could negatively impact customer satisfaction.
FoodHub should investigate the causes of these delays and work with
restaurants and delivery personnel to reduce long wait times.

### **Question 14:** The company wants to analyze the delivery time of the orders on weekdays and weekends. How does the mean delivery time vary during weekdays and weekends? [2 points]

In [20]:
# Get the mean delivery time on weekdays and print it
print('The mean delivery time on weekdays is around',
      round(df[df['day_of_the_week'] == 'Weekday']['delivery_time'].mean()),
     'minutes')

# Get the mean delivery time on weekends and print it
print('The mean delivery time on weekends is around',
      round(df[df['day_of_the_week'] == 'Weekend']['delivery_time'].mean()),
     'minutes')

The mean delivery time on weekdays is around 28 minutes
The mean delivery time on weekends is around 22 minutes


The mean delivery time on weekdays is approximately 28 minutes, compared
to 22 minutes on weekends. This difference is likely explained by heavier
traffic congestion on weekdays due to commuters, which slows down delivery
personnel. On weekends, lighter traffic allows delivery personnel to travel
more quickly between restaurants and customers. FoodHub should consider
accounting for traffic patterns when estimating delivery times, and may
want to communicate longer expected delivery windows to customers who
order on weekdays to better manage their expectations.

### Conclusion and Recommendations

### **Question 15:** What are your conclusions from the analysis? What recommendations would you like to share to help improve the business? (You can use cuisine type and feedback ratings to drive your business recommendations.) [5 points]

### Conclusions:

Based on the analysis of the FoodHub dataset, the following conclusions
can be drawn:

* American cuisine is the most popular cuisine on weekends, followed
  by Japanese, Italian, Chinese, and Mexican.
* Shake Shack, The Meatball Shop, Blue Ribbon Sushi, and Blue Ribbon
  Fried Chicken are the most frequently ordered from restaurants on
  the platform.
* 736 orders (38.8%) were not rated, representing a significant gap
  in customer feedback data.
* The four restaurants eligible for the promotional offer all have
  average ratings greater than 4 and more than 50 ratings: The Meatball
  Shop, Blue Ribbon Fried Chicken, Shake Shack, and Blue Ribbon Sushi.
* Weekday deliveries take longer than weekend deliveries (28 minutes
  vs 22 minutes), likely due to heavier traffic congestion on weekdays.

### Recommendations:

Based on the conclusions above, the following recommendations are made:

* FoodHub should recruit more American and Japanese restaurants to the
  platform to meet high weekend demand in these cuisine categories.
* The four restaurants with ratings greater than 4 should be featured
  prominently in the app and in marketing campaigns to drive more orders.
* FoodHub should introduce incentives such as discount vouchers or
  loyalty points to encourage customers to leave ratings, as more rating
  data would help identify high and low performing restaurants across
  all cuisine types.
* Restaurants with lower ratings should receive targeted feedback and
  support to improve food quality and preparation times, as customer
  ratings are directly tied to repeat business and platform reputation.

###AI Usage Statement:

I used Claude (claude.ai) by Anthropic to assist me with this assignment.
AI was used to help guide me through filling in the code blanks, interpreting
the results of each question, and drafting the written analysis in the text
cells. I verified all code outputs independently by running each cell in
Google Colab and confirming the results before accepting them. The prompts
I used included asking Claude to help me understand what each blank required,
requesting interpretations of the numerical results, and asking for help
structuring the conclusions and recommendations sections based on the
professor's instructions.